include/counter.h

```diff
@@ -22,35 +22,133 @@
+			  * dependent.
+			  */
+
+	bool absolute; /*!< If true ticks are treated as absolute value, else
+			*   it is relative to the counter reading performed
+			*   during the call.
+			*/
+
+	u32_t ticks; /*!< In case of absolute flag is set, maximal value that
+		      *   can be set equals wrap value (counter_get_wrap).
+		      *   Otherwise counter_get_max_relative_alarm() returns
+		      *   maximal value that can be set. If counter is clock
+		      *   driven then ticks can be converted to microseconds
+		      *   (see @ref counter_ticks_to_us). Alternatively,
+		      *   counter implementation may count asynchronous events.
+		      */
+
+	u32_t period; /*!< If non-zero alarm is repeated. If counter is clock
+		       *   driven then period can be converted to microseconds
+		       *   (see @ref counter_ticks_to_us). Alternatively,
+		       *   counter implementation may count asynchronous events.
+		       */
+
+	counter_alarm_callback_handler_t handler; /*!< handler called on alarm
+						   *   (cannot be NULL).
+						   */
+};
+
+struct counter_wrap_callback;
+
+typedef void (*counter_wrap_callback_handler_t)(struct device *dev,
+					struct counter_wrap_callback *cb);
+
+/** @brief Wrap callback structure.
+ *
+ * Used to set an overflow. *
+ * Beware such structure should not be allocated on stack.
+ */
+struct counter_wrap_callback {
+	u32_t ticks; /*!< If counter is clock driven then period can be
+		      *   converted to microseconds (see
+		      *   @ref counter_ticks_to_us). Alternatively, counter
+		      *   implementation may count asynchronous events.
+		      */
+	counter_wrap_callback_handler_t handler; /*!< handler called on wrapping
+						  *   (can be NULL).
+						  */
+};
+
+/** @brief Structure with generic counter features. */
+struct counter_config_info {
+	u32_t max_wrap; /*!< Maximal (default) wrap value on which counter is
+			 *   reset (cleared or reloaded).
+			 */
+
+	u32_t freq; /*!< Frequency of the source clock if synchronous events
+		     *   are counted.
+		     */
+
+	bool count_up; /*!< Flag indicating direction of the counter. If true
+			*   counter is counting up else counting down.
+			*/
+
+	u8_t channels; /*!< Number of channels that can be used for setting
+			*   alarm, see @ref counter_set_alarm.
+			*/
+
+	const void *config_info; /*!< Pointer to implementation specific
+				  *   configuration.
+				  */
+};
 
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
 typedef int (*counter_api_set_alarm)(struct device *dev,
-				     counter_callback_t callback,
-				     u32_t count, void *user_data);
+				     struct counter_alarm_callback *cb);
+typedef int (*counter_api_disable_alarm)(struct device *dev,
+				     struct counter_alarm_callback *cb);
+typedef int (*counter_api_set_wrap_alarm)(struct device *dev,
```

Comment:
What about timers which have fixed wrap value?

Reply:
then it will return -ENOTSUP

Reply:
But then it won't be possible to install a callback function for the wrap event. Maybe this fixed value should be used in such case?